<a href="https://colab.research.google.com/github/Omkar210/Statistics-and-ML/blob/main/Day17.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
# import warnings
# warnings.filterwarnings("ignore")
# pd.set_option('display.max_columns', None)
from numpy import random
from collections import Counter
from numpy.linalg import inv
from numpy.linalg import eig
import matplotlib
from matplotlib import pyplot as plt
import seaborn as sns
import pylab
from pylab import legend
from pylab import plot, show, title, xlabel, ylabel
import scipy
from scipy import stats
from scipy.stats import binom,poisson,norm,t,expon,f
from sklearn.model_selection import train_test_split
import statsmodels
from statsmodels import stats
from statsmodels.stats import weightstats as ssw
import statsmodels.api as sm
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import statsmodels.stats
from statsmodels.stats.proportion import proportions_ztest
from scipy.stats import chisquare
from scipy.stats import chi2_contingency
from statsmodels.stats import rates
from statsmodels.stats.rates import test_poisson
from statsmodels.stats.rates import test_poisson_2indep
from scipy.stats import chi2
from scipy.stats import f
import statsmodels.api as sm
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from category_encoders import BinaryEncoder
from sklearn.metrics import confusion_matrix, classification_report
import os
print(os.getcwd())
os.chdir('/content/drive/MyDrive/CDAC')
os.getcwd()

/content/drive/MyDrive/CDAC


'/content/drive/MyDrive/CDAC'

In [ ]:
df = pd.read_excel('CDAC_DataBook.xlsx', sheet_name = 'diabetes')

df =df[['Glucose','BloodPressure','Age','DietType','Outcome']]

x = df.drop('Outcome', axis=1)
y = df.Outcome

diet_dummy = pd.get_dummies(x.DietType, drop_first=True, prefix='Diet').astype(int)
x = x.drop('DietType', axis=1)
x = pd.concat([x, diet_dummy], axis=1)

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=20)

x_train = sm.add_constant(x_train, prepend=False)
mod3 = sm.Logit(y_train, x_train).fit()
print(mod3.summary())

Optimization terminated successfully.
         Current function value: 0.333614
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:                Outcome   No. Observations:                  614
Model:                          Logit   Df Residuals:                      608
Method:                           MLE   Df Model:                            5
Date:                Sat, 06 Jun 2026   Pseudo R-squ.:                  0.4848
Time:                        09:37:01   Log-Likelihood:                -204.84
converged:                       True   LL-Null:                       -397.59
Covariance Type:            nonrobust   LLR p-value:                 3.933e-81
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Glucose           0.0368      0.005      7.867      0.000       0.028       0.046
BloodPressure     0.

Precision :
$$
\text{Precision} = \frac{TP}{TP + FP}
$$

$$
\text{Precision} = \frac{TN}{TN + FN}
$$

Recall :
$$
\text{Recall} = \frac{TP}{TP + FN}
$$

$$
\text{Recall} = \frac{TN}{TN + FP}
$$

F1 - Score :
$$
F_1 = 2 \cdot \frac{\text{Precision} \cdot \text{Recall}}
{\text{Precision} + \text{Recall}}
$$

In [ ]:
x_test = sm.add_constant(x_test, prepend=False)
y_pred_prob = mod3.predict(x_test)
y_pred = []

for i in y_pred_prob:
  if i >= 0.5:
    y_pred.append(1)
  else:
    y_pred.append(0)

print('''True Negative     False Positive
False Negative    True Positive\n''')

print(f'Confusion Matrix: \n{confusion_matrix(y_test, y_pred)}\n\n')
print(f'Classification Report: \n{classification_report(y_test, y_pred)}')

True Negative     False Positive
False Negative    True Positive

Confusion Matrix: 
[[95  6]
 [13 40]]


Classification Report: 
              precision    recall  f1-score   support

           0       0.88      0.94      0.91       101
           1       0.87      0.75      0.81        53

    accuracy                           0.88       154
   macro avg       0.87      0.85      0.86       154
weighted avg       0.88      0.88      0.87       154



category data - R2

continuous data -  accuracy

In [ ]:
print((95+40) / (95+40+6+13))
135/154  # Accuracy

# Precision
Positive-decision = TP / (TP + FP)

Negative-decision = TN / (TN + FN)

# Recall
TP / (TP + FN)

TN / (TN + FP)

# F1 Score
2 * (Precision * Recall) / (Precision + Recall)

0.8766233766233766


0.8766233766233766

In [ ]:
mydata = pd.DataFrame([[100,75,50,1,0,1],[100,75,51,1,0,1]], columns = x_train.columns)

In [ ]:
mypreds = mod3.predict(mydata)
LOR_51 = np.log(0.069934/(1-0.069934))
LOR_50 = np.log(0.067899/(1-0.067899))
LOR_51-LOR_50  # coeff of Age

np.float64(0.03171625976306691)

This 0.0317 define: the coeff indicate when age increase/change by 1 unit the change in the log of odd's ratio in favor of getting deibetes.

In [ ]:
mydata = pd.DataFrame([[100,75,50,1,0,1],[100,75,50,0,0,1]], columns = x_train.columns)
mypreds = mod3.predict(mydata)
mypreds
LOR_51 = np.log(0.067899/(1-0.067899))
LOR_50 = np.log(0.787203/(1-0.787203))
LOR_51-LOR_50   # coeff of Diet_2

np.float64(-3.927567368079736)

# One Hot Encoding

- Is used for Nominal scale variable - no logical order
- Not useful when numner of categories are too large
- Regression, ANN, KNN, SVM

# Ordinal Encoding

- Variable with inheritance rankinh or rating
- preseve logical order


In [ ]:
df = pd.DataFrame({
    'Education':[
        'Graduate',
        'PG',
        'High School',
        'Graduate',
        'PG'
    ]
})

education_map = {
    'High School':1,
    'Graduate':2,
    'PG':3
}

df['Education_Encoded'] = (
    df['Education']
    .map(education_map)
)

print(df)

     Education  Education_Encoded
0     Graduate                  2
1           PG                  3
2  High School                  1
3     Graduate                  2
4           PG                  3


In [ ]:
# Frequency encoding
df = pd.DataFrame({
    'City':[
        'Mumbai',
        'Mumbai',
        'Mumbai',
        'Delhi',
        'Delhi',
        'Pune'
    ]
})

freq = df.City.value_counts()

df['City_Frequency'] = (df.City.map(freq))
df

,City,City_Frequency
0,Mumbai,3
1,Mumbai,3
2,Mumbai,3
3,Delhi,2
4,Delhi,2
5,Pune,1


In [ ]:
df = pd.DataFrame({
    'City': [
        'Mumbai',
        'Pune',
        'Delhi',
        'Chennai',
        'Kolkata'
    ]
})

encoder = BinaryEncoder(cols=['City'])
df_encoded = encoder.fit_transform(df)
print(df_encoded)

df_encoded = encoder.transform(df)
result = pd.concat([df, df_encoded], axis=1)
print(result)

   City_0  City_1  City_2
0       0       0       1
1       0       1       0
2       0       1       1
3       1       0       0
4       1       0       1
      City  City_0  City_1  City_2
0   Mumbai       0       0       1
1     Pune       0       1       0
2    Delhi       0       1       1
3  Chennai       1       0       0
4  Kolkata       1       0       1
